File: src/sim/rng.ts
export class Rng {
  private state: number

  constructor(seed = Date.now()) {
    this.state = seed >>> 0
  }

  next(): number {
    this.state = (1664525 * this.state + 1013904223) >>> 0
    return this.state / 0x100000000
  }

  int(min: number, maxExclusive: number): number {
    const span = maxExclusive - min
    return min + Math.floor(this.next() * span)
  }

  shuffle<T>(arr: T[]): void {
    for (let i = arr.length - 1; i > 0; i -= 1) {
      const j = this.int(0, i + 1)
      ;[arr[i], arr[j]] = [arr[j], arr[i]]
    }
  }
}

File: go-sim/internal/sim/rng.go
package sim

type RNG struct {
	state uint32
}

func NewRNG(seed uint32) *RNG {
	if seed == 0 {
		seed = 1
	}
	return &RNG{state: seed}
}

func (r *RNG) Next() float64 {
	r.state = 1664525*r.state + 1013904223
	return float64(r.state) / float64(uint64(1)<<32)
}

func (r *RNG) Int(min, maxExclusive int) int {
	if maxExclusive <= min {
		return min
	}
	span := maxExclusive - min
	return min + int(r.Next()*float64(span))
}

func (r *RNG) Shuffle(ids []int) {
	for i := len(ids) - 1; i > 0; i-- {
		j := r.Int(0, i+1)
		ids[i], ids[j] = ids[j], ids[i]
	}
}
